Analyses/src/PhysVolumeInfoPrinter_module.cc

```diff
@@ -0,0 +1,72 @@
+// Andrei Gaponenko, 2021
```

Comment:
or add anything missing to Print/src/PhysicalVolumePrinter.cc

Reply:
Hi Ray,  I did not realize there was Print/src/PhysicalVolumePrinter.cc
I removed the new module in question from Analyses.

However I want to understand the choice of using a single module for
printing unrelated data products.  I only see disadvantages here, what
am I missing?

- My major use case for printing code is for code development.  When I
  start working on a new feature, parts of Offline may be broken by
  the initial change.  At first I want to be able to build the minimal
  code set to test the concept and decide whether to fix the rest of the
  code or abandon the initial change and try something else instead.
  With this mindset, a standalone volume printer like
  PhysVolumeInfoPrinter_module.cc would work as long as
  PhysicalVolumeInfoMultiCollection builds, regardless of the rest
  of the code.  On the other hand PrintModule will not be usable
  unless all the crv/calo/etc classes build.

- There is conceptual overhead: instead of using the art framework we
  invent a "printing framework" with its own steering.  Instead of
  adding plugin product-printing modules to an art end path, one has
  to know to use the custom "verbosity" setting to enable or disable
  printouts.  (And one can define a fhicl sequence with "all known
  types" to handle the "print-everything" use case.)

- There is coding overhead: instead of a single module file per data
  product one has to use two file (header+cc), plus the main module
  file.  

   $ wc -l Print/inc/GenParticlePrinter.hh Print/src/GenParticlePrinter.cc 
     46 Print/inc/GenParticlePrinter.hh
     99 Print/src/GenParticlePrinter.cc
    145 total
   
   $ wc -l Analyses/src/PrintGenParticles_module.cc 
   62 Analyses/src/PrintGenParticles_module.cc

Comments?

Reply:
Sorry for the delay, I didn't see the comment

On bullet one, I'm not sure I understand what you are doing.  You are inserting new code into a full simulation sequence, while selectively removing certain SConscripts to avoid building certain subdirectories?  Or do you delete most modules from your area then restore them later?  If so, this seems like a pretty radical way of working, but I can see in some cases it might be smart.  Here I'd just throw it out to the group - is it important to support this use case?  I'm also not sure if it is less restrictive to have 36 modules in Analysis that need to be linked to various products, or one Print subdirectory that needs to be linked to al products.  Unless you are deleting modules.  Or maybe I'm missing the point of how you propose to organize the code.

On bullet two, I don't know if it is any more confusing or harder to deal with 36 modules rather than 36 verbosity switches in one fcl.  Again I'd just throw it out to the group, I prefer my approach of course, but if the majority doesn't, I can accept that.  One point to keep in mind is that the printer classes are designed to be used anywhere, and be called with any form of handle or product reference the user might have on hand.  So I imagined users would include the printer classes into their analysis code to help with debugging, like dumping only strange events. I like this feature but I don't think many people use it.  If you were to keep this feature by including one print class in one module, I'm not sure where the code goes and how it helps with bullet one.

on bullet three, I'd only mention the feature I mentioned in the previous paragraph.

So I don't have strong defenses for the current approach, other than it is already there.  I also don't use your style and don't see a big advantage to it, but I'm probably missing at least one point.  So bottom line, I'm willing to reorganize if the group decides it would be a good idea.  A good next step would be to show a slide at the meeting.

Reply:
> On bullet one, I'm not sure I understand what you are doing. 

Using the "-k" build switch to not break on the first error.  Then using a custom fcl to test a piece of new code in a minimalistic setting that still works if irrelevant modules fail to build.

> I'm also not sure if it is less restrictive to have 36 modules in Analysis that need to be linked to various products, or one Print subdirectory that needs to be linked to al products.

I do not care if 35 print modules did not compile when I am only interested in running the one print module I need at the moment.  In the monolithic design there is only one module, and I'd have to fix 36x more code to make it happy.

Andrei

Reply:
OK, I see how you are doing a partial build and why this matters.  In your style, it is important that the print code is only in the module, not a library, so that would seem to block my feature of being able to include the print routine anywhere.  Also the Print library only depends on a few other libraries, most notably the Products libraries, so even if you have half the build failing, it is still likely that the Print library builds.  It also doesn't care if there are modules that access the old product and fail.

I don't think the case is obvious, so I'd go by the consensus - if the group decides it is important to rework the Print library, I can do that.